uber/models.py

```diff
@@ -1625,6 +1817,464 @@ def must_contact(self):
             sorted({'({}) {}'.format(dept_names[dept], ' / '.join(chairs[dept])) for dept in locations})))
 
 
+class PromoCodeWord(MagModel):
+    """
+    Words used to generate promo codes.
+
+    Attributes:
+        word (str): The text of this promo code word.
+        normalized_word (str): A normalized version of `word`, suitable for
+            database queries.
+        part_of_speech (int): The part of speech that `word` is.
+            Valid values are:
+
+            * 0 `ADJECTIVE`: `word` is an adjective
+
+            * 1 `NOUN`: `word` is a noun
+
+            * 2 `VERB`: `word` is a verb
+
+            * 3 `ADVERB`: `word` is an adverb
+
+        part_of_speech_display (str): A human readable description of
+            `part_of_speech`.
+    """
+
+    ADJECTIVE = 0
+    NOUN = 1
+    VERB = 2
+    ADVERB = 3
+    PART_OF_SPEECH_OPTS = [
+        (ADJECTIVE, 'adjective'),
+        (NOUN, 'noun'),
+        (VERB, 'verb'),
+        (ADVERB, 'adverb')]
+    PARTS_OF_SPEECH = dict(PART_OF_SPEECH_OPTS)
+
+    word = Column(UnicodeText)
+    part_of_speech = Column(Choice(PART_OF_SPEECH_OPTS), default=ADJECTIVE)
+
+    __table_args__ = (
+        Index('uq_promo_code_word_normalized_word_part_of_speech',
+            func.lower(func.trim(word)), part_of_speech, unique=True),
+        CheckConstraint(func.trim(word) != '',
+            name='ck_promo_code_word_non_empty_word'))
+
+    _repr_attr_names = ('word',)
+
+    @hybrid_property
+    def normalized_word(self):
+        return self.normalize_word(self.word)
+
+    @normalized_word.expression
+    def normalized_word(cls):
+        return func.lower(func.trim(cls.word))
+
+    @property
+    def part_of_speech_display(self):
+        return self.PARTS_OF_SPEECH[self.part_of_speech].title()
+
+    @presave_adjustment
+    def _attribute_adjustments(self):
+        # Replace multiple whitespace characters with a single space
+        self.word = re.sub(r'\s+', ' ', self.word.strip())
+
+    @classmethod
+    def group_by_parts_of_speech(cls, words):
```

Comment:
It's totally not a language generator. Just broad categories of word types to help organize promo code words.